### ISSUE TITLE
Android app does not check user trusted self-signed certificates

### ISSUE THREAD

**Issue by atb00ker** [tailscale/tailscale#8085]
### What is the issue?

I am a freeloader home user who decided to host the service myself using `headscale` to not stress your systems without paying.
Headscale is working fine on my linux boxes using tailscale client.
However,
I installed the certificate on my phone and checked if it's working using firefox; and the certificate is installed properly.
However, when I try to connect to the control plane using the android app, it still complains "remote error: tls: bad certificate".

I suspect that the app is only looking at the system certificates instead of all / user certificates.

P.S: I understand that this will be lowest priority, but I'd appreciate someone helping out with this, cheers and thanks for letting us use other control planes.

### Steps to reproduce

1. Create a control plane with self-signed certificates
2. Add the certificate's CA as trusted in android settings.
3. Try to connect to the control plane
4. Notice the certificates are still not trusted. 😭 

### OS

Android

### OS version

Android 11

### Tailscale version

1.38.2

**Comment by atb00ker** [comment]
P.S: I am not an android developer, but if the solution is as easy as https://stackoverflow.com/a/72280826/6410464
and you folks don't mind this change, I'll try to give it a go! 😄

**Comment by DentonGentry** [comment]
> decided to host the service myself using headscale to not stress your systems without paying.

As other people will read that statement I do want to address it: we want people to use the service, for free. Tailscale is able to do that because the incremental cost of adding a free user is very low, and those free users in turn provide the word-of-mouth growth which leads some percentage to become paying — mostly corporate — users. Our CEO wrote a blog post about it: [How our free plan stays free](https://tailscale.com/blog/free-plan/) 

You're [welcome to run a Headscale server](https://tailscale.com/blog/opensource/), if you have reasons to want to run it, but please don't feel compelled to. The Tailscale service is structured to be free for personal use.

--------

Newer Android releases make it harder to install user certificates because of the history of malware abuse, apps tricking users into installing a certificate which is then leveraged in a later step in an attack.

I'm not very enthused about having the Tailscale Android app trust those CAs, since almost no Android apps do so. I'm concerned that trusting those certificates would become a line in a future security report as a step in an exploit.

--------

Surely you're familiar with LetsEncrypt, which would produce a valid certificate. Is there a reason to prefer a self-signed certificate?

**Comment by hnrd** [comment]
> Surely you're familiar with LetsEncrypt, which would produce a valid certificate. Is there a reason to prefer a self-signed certificate?

Public certificates get added to certificate transparency logs (as they should). Using an internal CA or self-signed certificates for internal services keeps those private.

I don't share your observation about almost no android apps honouring the system CA store, I seldomly see apps that don't. The biggest app that doesn't is probably firefox/fennec, but it has a (hidden) configuration option to enable the system CA store. Maybe that is a way for tailscale, too? Disable system CA until the user explicitly enables it?

**Comment by MatheusJCastro** [comment]
This is a feature that would be great to have. I have a residential internet provider and they block access through most common ports like 80 and 443, so I can't use acme challenges. Moreover I use a free DNS provider, which doesn't let me write TXT records. Because of this, I have my own CA and have it installed on all my devices.

Tailscale Android app is the first software that doesn't let me use self-signed certificates. However other tailscale clients that I use (linux and windows) work fine.

Security concerns shouldn't restrict a feature based on what users **may or may not** do (in this case install malicious self-signed certificates). If that was the case we should close the internet itself. Besides that, as [pointed out here](https://github.com/tailscale/tailscale/issues/8085#issuecomment-1537430052), Android made it more difficult to install these certificates to prevent this type of attack, but they didn't block it at all, giving the freedom of choice for users.

If this is still a security concern, a good implementation would be the same as Firefox, to explicitly activate self-signed certs in the app and maybe even as a hidden feature.

> the solution is as easy as https://stackoverflow.com/a/72280826/6410464

I'm also not an android developer, but I git cloned the android repository and also made the modifications pointed out here, compiled and installed in my device. However without success.

**Comment by atb00ker** [comment]
While I personally disagree with the "security reasons" wall for not allowing self-signed certificates,
I've met the same wall with some other android projects as well, example, onlyoffice doesn't support my self-signed nextcloud drive on android; works fine everywhere else.
It's been quite a work talking about this with different projects and it's clear while some projects are happy to add the support, other projects are reluctant to do the same. It seems to be a common belief in the android community that self-signed certificates are insecure.

My long term plan is to try creating a fork that works or hire someone on fiverr to help me create the fork but that task has taken a backseat.
If someone can help with the first time process of making the changes and making a working built (as I lack android development knowledge); I can help maintain that project and keep the fork updated and release in sync with the official app so that we, the few, can use these tools and update via Obtainium.

**Comment by gloowa** [comment]
Hi, I'm just going to chime in here - possibility of having a tailscale andoid client trust a user-defined root ca would be much appreciated. I've just managed to get my headscale configured after 3 days of fighting with docker, nginx and hundreds of small gotcha's, and now i've crashed into this issue, making the entire effort worthless since the entire point of the excercise was to allow access to my home network from my smartphone. 

ZeroSSL only gives certificates out for 3 months, and I'd rather not have to go through the entire hoopla of re-configuring my nginx just to re-verfy domain ownership, to have the cert renewed.

If anyone knows a trusted org that will issue long term domain ssl certs for free, dm me.

**Comment by ghost** [comment]
Chiming in here too, with the same hope of getting a self signed to work. Like the last reply here before me, I too spent a lot of time learning how to set up Headscale with reverse proxy and learning how headers and 500-errors work, only to then learn that I cannot connect with this app without a publicly signed certificate. 

Because of everything that is happening around us in terms of privacy, laws and legislations, we do not know what the near future holds in terms of internet as we know it.. and in such an instance where rules could change on a dime (see Google's "Apple-fication" of walled garden), it would be great to give us the option to FULLY self host. That's the spirit of open source, to make it open and free.. Ofcourse I am free to modify this myself, allthough some of us lack the knowledge on how to do so.. 

About to try out the git-clone thing, but I see in the comments here that it didn't work

**Comment by ghost** [comment]
-Alright I've spent two days trying to find a fix. Tried the xml thing, that didn't work as the app wouldn't want to install without signing.
-Signing it witl a testkey didn't work either
-modifying .go files and trying to bake in a config and make it bypass tlsConfig and inject a new config that points to selfsigned certificate was a fun task, but I get constant "go bind" errors because apparently Tailscale has it's own custom version of this, and it makes it impossible to rebuild this app.

**Comment by atb00ker** [comment]
Thanks for trying @groke76 ! 😄 
Just to share an update, my imperfect solution has been installing tailscale on an OpenWRT router and calling it a day.
Of course that means I need to be connect to that router but other than that... it works!

**Comment by ghost** [comment]
> Thanks for trying [@groke76](https://github.com/groke76) ! 😄 Just to share an update, my imperfect solution has been installing tailscale on an OpenWRT router and calling it a day. Of course that means I need to be connect to that router but other than that... it works!

Yeah i have the same app on my router. I initially wanted to join it to an external private VPN over Hwadscale, with personal DERP server and totally free from any third party, and be able to rpute access as I wish, bit this tiny bit lack of user CA stops that unfortunately.

Tailscale on the router itself, without any Headscale configuration, does it give any benefit over just normal WireGuard?

**Comment by YCJDWCJI** [comment]
self-signed cert are now supported onAndroid, see [#11776] .

[#11776]: https://github.com/tailscale/tailscale/issues/11776#issuecomment-2520955317

**Comment by ghost** [comment]
Thats great!! It says it was posted las year, though. Did we miss something in this thread? 🤔

**Comment by LoganRupe** [comment]
Is there any update on the use of custom/user-installed CA with tailscale client/app on android devices?

**Comment by LoganRupe** [comment]
Hey all — I've put up a PR that fixes this: tailscale/tailscale-android#757

The root cause is that Go's `crypto/x509` on Android only reads system CAs from `/system/etc/security/cacerts/` and doesn't consult the Android user trust store at all. The fix bridges user-installed CA certs from Android's `KeyStore` into Go's TLS stack at startup via `SSL_CERT_DIR`, without touching the system cert path.

It's been tested successfully against a Headscale instance using a self-signed CA. Happy to make any changes if there are concerns — would love to get some eyes on it.

**Comment by bradfitz** [comment]
@LoganRupe, I sent https://github.com/tailscale/tailscale/pull/19280 which I think would make https://github.com/tailscale/tailscale-android/pull/757 be simpler, cleaner, and more robust to future changes.

**Comment by LoganRupe** [comment]
> [@LoganRupe](https://github.com/LoganRupe), I sent [#19280](https://github.com/tailscale/tailscale/pull/19280) which I think would make [tailscale/tailscale-android#757](https://github.com/tailscale/tailscale-android/pull/757) be simpler, cleaner, and more robust to future changes.

Thanks for picking this up @bradfitz — I've opened tailscale/tailscale-android#775 which implements the Android side of your ExtraRootCAs approach. It reads user-installed CA certs from AndroidCAStore via gomobile, builds an *x509.CertPool from x509.SystemCertPool() + the user certs, and assigns it to sys.ExtraRootCAs before startup. Closes my earlier tailscale/tailscale-android#757.

**Comment by bradfitz** [comment]
@LoganRupe's PR is merged. Thanks!
